[staging] {nixos/}systemd: enable BTF; enable {nsresourced,mountfsd}

msanft opened the pull request:
<!--
^ Please summarise the changes you have done and explain why they are necessary here ^

For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

Re-target of #404715 against staging.

This adds support for rootless systemd-nspawn containers on NixOS by doing a few things:
- Changing the `linuxManualConfig` builder by unconditionally emitting a `vmlinux` ELF in the `dev` output, not just when `CONFIG_MODULES=y` (also removing an IFD requirement for that).
- Adding BTF information for the systemd package's BPF programs, which are a prerequisite for `nsresourced` and `mountfsd`. This is achieved by building a minimal custom kernel, generating a `vmlinux.h` file from that, and consuming that in systemd. This does not create a strong dependency between the systemd package and a specific kernel version, as 
the header file is only used for initial compilation, and BPF symbols are being relocated at runtime later on using BPF CO-RE.
- Adding toggles for the `systemd-{nsresourced,mountfsd}` binaries in the systemd package. These are passed-through, to be consumed by the NixOS module.
- Conditionally enabling the aforementioned units in the NixOS module for systemd.

Change can be quickly verified by the following pre-change and post-change:
```sh
sudo $(nix build .#systemd.out --print-out-paths --no-link)/lib/systemd/systemd-nsresourced
```

Prior to the change, this will show:
```
Not setting up BPF subsystem, as functionality has been disabled at compile time.
```

After the change:
```
userns-restrict BPF-LSM program userns_restrict_path_chown attached.
userns-restrict BPF-LSM program userns_restrict_path_mkdir attached.
userns-restrict BPF-LSM program userns_restrict_path_mknod attached.
userns-restrict BPF-LSM program userns_restrict_path_symlink attached.
userns-restrict BPF-LSM program userns_restrict_path_link attached.
userns-restrict BPF-LSM program userns_restrict_free_user_ns attached.
```

## Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [x] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [ ] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [25.05 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) (or backporting [24.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) and [25.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2505.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#reviewing-contributions
-->

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

dramforever commented on pkgs/os-specific/linux/kernel/manual-config.nix:
```diff
@@ -427,77 +426,80 @@ lib.makeOverridable (
+        postInstall =
+          ''
+            mkdir -p $dev
+            cp vmlinux $dev/vmlinux
```
```suggestion
            cp vmlinux $dev/
```

Saves one unnecessary change and rebuild on linux when `isModular`

dramforever commented on pkgs/os-specific/linux/systemd/vmlinux-btf.nix:
```diff
@@ -0,0 +1,62 @@
+      makeTarget = "allnoconfig";
+    }).overrideAttrs
+      (oldAttrs: {
+        depsBuildBuild = oldAttrs.depsBuildBuild ++ [ pahole ];
```
Can we just add this to `linuxConfig`? It just seems like a footgun for custom config users otherwise. I wonder what other tools we're missing there...

ElvishJerricco commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -162,6 +167,8 @@
   withUserDb ? true,
   withUtmp ? !stdenv.hostPlatform.isMusl,
   withVmspawn ? true,
+  withNsresourced ? true,
+  withMountfsd ? true,
```
These need to be disabled by `systemdMinimal`.

ElvishJerricco commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -543,6 +550,11 @@ stdenv.mkDerivation (finalAttrs: {
 
+      # BPF
+      (lib.mesonEnable "bpf-framework" withLibBPF)
+      (lib.mesonOption "vmlinux-h" "provided")
+      (lib.mesonOption "vmlinux-h-path" vmlinux-h)
```
These should not be provided when `withLibBPF = false`, right? Otherwise `vmlinux-h` will be built as an input when it isn't even used.

ElvishJerricco commented on pkgs/os-specific/linux/systemd/vmlinux-btf.nix:
```diff
@@ -0,0 +1,59 @@
+    (linuxConfig {
+      inherit (linuxHeaders) src;
+      makeTarget = "allnoconfig";
+    }).overrideAttrs
+      (oldAttrs: {
+        buildPhase =
+          oldAttrs.buildPhase
```
This is mostly curiosity, but could this have been done by providing `extraSrtructuredConfig` via `kernelPatches` rather than overriding the build phase like this?

msanft replied:
It's very unfortunate, and I have tried each and every other option to make this work, but `linuxConfig` doesn't have a convenient way to pass patches to the kernel config like `buildLinux` and friends have.

ElvishJerricco commented on pkgs/top-level/linux-kernels.nix:
```diff
@@ -865,6 +865,9 @@ in
         ++ lib.optionals (lib.versionAtLeast version "4.16") [
           buildPackages.bison
           buildPackages.flex
+        ]
+        ++ lib.optionals (lib.versionAtLeast version "5.10") [
+          buildPackages.pahole
```
Can you explain this addition?

msanft replied:
Since 5.18 (typo, will correct), pahole is an optional dependency for building a kernel config: https://cateee.net/lkddb/web-lkddb/PAHOLE_VERSION.html

alyssais replied:
Should that go in a separate commit with an explanatory commit message then?

ElvishJerricco commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -87,6 +88,10 @@
 
+  # vmlinux.h file containing the kernel's BTF information,
+  # used throughout systemd's BPF programs.
+  vmlinux-h ? "${callPackage ./vmlinux-btf.nix { }}/vmlinux.h",
```
I think I would like to see this added to the passthru if only so that it's easy to build it directly or read its build log.

ElvishJerricco replied:
Oh, it's a thin `runCommand` wrapper around the actual `vmlinux` derivation. I guess that should be in the `runCommand`'s `passthru` as well.

alyssais commented on pkgs/os-specific/linux/kernel/manual-config.nix:
```diff
@@ -427,77 +426,80 @@ lib.makeOverridable (
         # which trips the broken symlink check. So, just skip it. We'll know if it explodes.
         dontCheckForBrokenSymlinks = true;
 
-        postInstall = optionalString isModular ''
```
Have you researched why this was previously only done for modular builds?

dramforever replied:
My understanding is that `$dev/vmlinux` was added in 53a45825523566b8c9c5ff135e8ff75e05558d85 and that broke non-`isModular` builds because the dev output is conditional on `isModular`. 7ce1f0b8ae74fb7dbaa02bbcbe856026314b8c95 chose to fix this (IIUC, I don't see any specific discussion on the build error) by making `$dev/vmlinux` also conditional, whereas making the dev output unconditional would also work

alyssais replied:
Okay, makes sense. Would be good to note this in the commit message.

alyssais commented on pkgs/os-specific/linux/systemd/default.nix:
```diff
@@ -87,6 +88,10 @@
 
+  # vmlinux.h file containing the kernel's BTF information,
+  # used throughout systemd's BPF programs.
+  vmlinux-h ? "${callPackage ./vmlinux-btf.nix { }}/vmlinux.h",
```
We might end up needing this for packages that use bpf CO-RE other than systemd, right? Why not make it a top-level attribute?

dramforever replied:
The configs selected are systemd-specific. This is also the reason https://github.com/libbpf/vmlinux.h wouldn't work --- it doesn't have enough configs enabled.

Although... I guess it can't really hurt to have one central kitchen sink `vmlinux.h` in Nixpkgs. What do you think?

msanft replied:
I don't think this would need to happen in this PR? These files are inherently use-case-specific.

alyssais replied:
> Although... I guess it can't really hurt to have one central kitchen sink `vmlinux.h` in Nixpkgs. What do you think?

I think this would make much more sense.

rnhmjoj replied:
- Sorry for the bad naming in the example: it builds the vmlinux ELF. I needed vmlinux to generate vmlinux.h later, with `bpftool`. Same as you do, I think.

- Maybe, but at least a user can just supply a kernel binary, rather than the sources to generate this infos. It seems more convenient to me.

alyssais commented on pkgs/os-specific/linux/systemd/vmlinux-btf.nix:
```diff
@@ -0,0 +1,59 @@
+    (linuxConfig {
+      inherit (linuxHeaders) src;
+      makeTarget = "allnoconfig";
+    }).overrideAttrs
```
I don't think we should be using overrideAttrs like this in Nixpkgs — we control the linuxConfig API, so we can add a proper parameter to it rather than doing a hacky override.

msanft replied:
I switched to `overrideAttrs` after getting [this review](https://github.com/NixOS/nixpkgs/pull/404715#discussion_r2076369685).

dramforever replied:
iiuc alyssa's idea was not to go back to plain `stdenv.mkDerivation`, but to add this as an argument to `linuxConfig`

msanft replied:
Alright, so add a generic way to add extra configuration options to `linuxConfig`?

msanft committed:
mergeKernelConfigs: init

This adds a generic builder for merging multiple kernel configurations.

msanft committed:
linuxManualConfig: add vmlinux to dev output

This adds the `vmlinux` ELF file to the `dev` output of the
`linuxManualConfig` derivation unconditionally, making it easier to
access the kernel ELF file without creating a modular kernel, which
previously has been the requirement for the `dev` output.

This previous decision was made because `$dev/vmlinux` was added in
53a4582 and broke non-`isModular` builds because the `dev` output is
conditional on `isModular`. 7ce1f0b wanted to fix this, and did so by
(unnecessarily) making `$dev/vmlinux` also conditional, even though
making the entire `dev` output uncoditional would also work.

msanft committed:
systemd: compile with BTF information

This adds the necessary BTF information for systemd's BPF programs. This relies on a purpose-built, minimal kernel, from which the `vmlinux.h` header file with the kernel-internal definitions is generated. Note that the systemd derivation is not depending on that specific kernel, it's just used for initial compilation. At runtime, the symbols are relocated using BPF CO-RE.

msanft committed:
nixos/systemd: enable {nsresourced,mountfsd}

This conditionally enables the nsresourced and mountfsd services, which are used to enable rootless systemd-nspawn containers.

alyssais commented on pkgs/top-level/all-packages.nix:
```diff
@@ -11132,7 +11132,12 @@ with pkgs;
+  inherit (linuxKernel)
+    buildLinux
+    linuxConfig
+    mergeConfigs
```
I think `mergeConfigs` is too generic as a top-level name.

alyssais commented on pkgs/top-level/linux-kernels.nix:
```diff
@@ -884,5 +892,28 @@ in
+      name ? "merged.config",
+      version ? (builtins.parseDrvName src.name).version,
+      a,
+      b,
```
IMO it would either make more sense to take two positional arguments, or ideally a list that could contain more than two.

alyssais commented on pkgs/os-specific/linux/systemd/vmlinux-btf.nix:
```diff
@@ -0,0 +1,62 @@
+      modDirVersion = lib.versions.pad 3 linuxHeaders.version;
+      allowImportFromDerivation = false;
+    }
+  );
```
No need for these parens.

alyssais commented on pkgs/os-specific/linux/systemd/vmlinux-btf.nix:
```diff
@@ -0,0 +1,62 @@
+      inherit (linuxHeaders) version src;
+      inherit configfile;
+      modDirVersion = lib.versions.pad 3 linuxHeaders.version;
+      allowImportFromDerivation = false;
```
Aren't the preceding two lines both the defaults?

alyssais commented on pkgs/os-specific/linux/systemd/vmlinux-btf.nix:
```diff
@@ -0,0 +1,62 @@
+in
+runCommand "systemd-btf"
+  {
+    version = linuxHeaders.version;
```
[nit] could be an inherit.

pwaller commented:
I've tested this out and it appears to mostly work as expected, so I think this is ready to be merged.

For anyone interested in developing a rootless nspawn flow: I tried using it for rootless containers but the requirements are pretty restrictive: it still requires authentication if the image you're using isn't cryptographically signed by a trusted key for example. I had to put the image in `/var/lib/machines` and give my user permission to write there (so that nspawn could write a lockfile there). 

I got pretty far down the road, but then faild at `Failed to take /etc/passwd lock: Permission denied`. I think this is coming from the chroot once the image is mounted. I've built a DDI image using repart. Not sure what needs fixing next, in case anyone else thinking about this.

@msanft (or anyone else), do you have a working rootless nspawn config to try out?

alyssais commented:
@emilazy @ElvishJerricco can you summarise the outcome of the discussion about this on Matrix last night?

emilazy commented:
1. Not clear how to provide a universal `vmlinux.h` that works well across runtime kernel configurations.
2. `libbpf/vmlinux.h` explicitly disclaims that it is for development only and has no stability guarantees.
3. The BPF code here only needs a UAPI header or two and a small handful of stub `struct` definitions with one or two fields each.
4. Also you need to do manual declarations to support e.g. multiple kernel versions after they change around fields anyway.
5. So it seems like the “prod‐ready” thing is to not rely on `vmlinux.h` and manually declare the stuff you need that isn’t stable UAPI.
6. We should make a minimal patch to inline the necessary declarations into the code and try to upstream it with an expanded version of the above rationale. This would also reduce some build system complexity they have. If they say yes, the problem is solved.
7. If they say no, back to square one. In that case probably the most elegant thing is to generate our own `vmlinux.h` files in `linux.dev` and just cross our fingers that it will work when used on a different kernel. This is the equivalent of what Arch does.

I will try to look at finishing the patch and sending a PR upstream today. @dramforever got it down to https://gist.github.com/dramforever/2e7c3f0a485ba8a45facaec991d81eb1 but I think much of the stuff in there should be unnecessary with the right UAPI includes.

Skyb0rg007 commented:
No worries! I just wanted to make sure the status and difficulties were documented to not discourage others from attempting a solution.

Ericson2314 commented:
I agree that it is untenable to put a kernel build extremely early (mass rebuild early) in the bootstrap order. Since systemd insists on not providing a modular build (e.g. using Meson subprojects, like Nix does), they are forcing us into this situation. There are only two ways out:

1. @dramforever's https://github.com/systemd/systemd/pull/40482
2. Build the `vmlinux.h`-using BPF-using components (`systemd-networkd`, `systemd-nsresourced`) as separate packages not part of the main `systemd`, but linking the main `systemd` (unlike `systemd`vs `systemdMinimal`, which is a lousy situation)

I would be happy with either solution --- 2 also works because then I don't believe `systemd-networkd` or `systemd-nsresourced` would be depended-upon enough to cause mass rebuilds, so a gratuitous kernel build dep just for those isn't *so* bad.

However, if upstream doesn't like either (they appear to not like 1, and my understanding is that they have always been against 2), then we should definitely go with 1, because it is a patch small enough that we could just land it ourselves downstream, over their objections, whereas 2 will be effectively impossible to maintain downstream.

dramforever commented:
@Ericson2314  We have option

3. Maintain a vmlinux.h subset, against upstream advice, but following [many other projects](https://github.com/search?q=path%3A%2F%5C%2Fvmlinux.h%24%2F&type=code) (Not all search results. Some digging required.)

I still believe this is not practically worse for compat than using the `vmlinux.h` of some random kernel version. It's *gonna* break, and require handling in the user code itself. See: https://github.com/systemd/systemd/pull/40482#issuecomment-3805991278 (Yes this is an "I believe I know better than systemd maintainers, and every distro that followed along with `-Dvmlinux-h-path`" situation)

`${kernel.dev}/vmlinux`-generated `vmlinux.h` is not a magic sauce that's somehow "the safest we can do".

One notable example IMHO is *perf* which switched to doing a `vmlinux.h` subset in 2023 (https://github.com/torvalds/linux/commit/a887466562b425bd6183bf75b523c1477c9fd22d) allowing it to be kernel version independent (see also https://github.com/NixOS/nixpkgs/pull/439229). Of course perf being in-tree is itself a quite unique situation...

Ericson2314 commented:
Yeah I would say

- applying your patch
- including a specially crafted `vmlinux.h` that contains exactly what those two BPF programs need and nothing more

are pretty isomorphic options ;)

I always would definitely share https://github.com/torvalds/linux/commit/a887466562b425bd6183bf75b523c1477c9fd22d in the PR thread. IMO in-tree BPF using a subset is a coup de grace. Even the kernel devs think the kernel is too heavy a debt! Even for code which is (effectively, because in tree) versioned the same as the kernel!

msanft commented:
> Also, here is a branch where I've written up an entire `vmlinux.h` for systemd. It seems to load eBPF LSMs fine but I haven't tried unpriv containers on it https://github.com/dramforever/nixpkgs/commits/systemd-with-vmlinux-h

I did the same at some point (perhaps even on this PR?) and it worked. For me, this would probably be the favorable way to move forward at the moment. This puts a burden on persons packaging BPF-integrated packages (e.g. systemd, psc), but I think it's the best we can do right now to not put kernel builds in the dependency chain of those packages.

P.S.: I agree with your analysis of CO-RE above.

RaitoBezarius commented:
[Summarizing a conversation that happened in NixOS systemd on Matrix](https://matrix.to/#/%21SsuQHVzf59iBgk6YIm%3A0upti.me/%24FYSzc-rTrV40R1Xjn02PoR81Vjmoz0qEi_0Z_7pSkA4?via=nixos.org&via=matrix.org&via=tchncs.de):

systemd upstream will not change their mind as mentioned above. After talking and looking at the options again with those who raised a concern, here are the current conclusions.

@emilazy @dramforever @alyssais accepts the compromise of either of these options:

- `inherit (linuxPackages_latest) ebpf-headers` and feeding these headers to systemd. This induces an extra kernel rebuild for bootstrapping and cross compilation to the latest Linux kernel version in the tree.
- `linuxPackages.systemd` (i.e. a systemd per kernel version) and feeding the headers induced by `linuxPackages.kernel` to that systemd. This induces `#nr of kernel in tree` extra rebuilds of systemd which are known [to take 6 minutes](https://hydra.nixos.org/build/319823096). The output of systemd is not a particularly large one neither.

Important notes:

- systemd will not make BPF a dependency in `libsystemd`, there's no cyclical dependency concern in the future.
- systemd will introduce CO-RE queries once a breakage will occur but will not deal with it until it's proven to be an actual problem. Other projects' problems are not considered valid because systemd does not have the same usage.
- Upgrading systemd at runtime seems still supported for now, but if we move it under `linuxPackages.systemd`, there's a possible miscommunication about it being able to upgrade.

The last solution seems preferred because it save up rebuilds for people who may be doing things with limited resources and pushes the (limited) churn on Hydra [except for the fact that there's 100 consumers of systemd in nixpkgs](https://github.com/search?q=repo%3ANixOS%2Fnixpkgs+%22systemd%2C%22+path%3Apkgs&type=code) which may just want libsystemd actually.

msanft commented:
I haven't been super active on Matrix recently, but if I interpret the recent comments here right, it seems that people on Matrix did (more or less) agree on `linuxPackages.systemd`, right? If so, I'd be happy to wait for the `libsystemd` treewide and then get back to work on this PR. Or does anyone have another idea on what I could do here to get this thing moving faster?

P.S.:

> systemd will introduce CO-RE queries once a breakage will occur but will not deal with it until it's proven to be an actual problem. Other projects' problems are not considered valid because systemd does not have the same usage.

If this is a statement from the systemd maintainers, the "hand-writing a minimal `vmlinux.h` for systemd" might be worth pursuing? I believe someone here PoCed that already. People did complain, pointing to the maintenance effort. But the systemd statement effectively implies that they will take care of the maintenance effort by introducing CO-RE when needed?
